metadata-ingestion/src/datahub/ingestion/source/metabase.py

```diff
@@ -0,0 +1,407 @@
+    options: Dict = {}
+    env: str = builder.DEFAULT_ENV
+
+    @validator("connect_uri")
+    def remove_trailing_slash(cls, v):
+        return config_clean.remove_trailing_slashes(v)
+
+
+class MetabaseSource(Source):
+    config: MetabaseConfig
+    report: SourceReport
+    platform = "metabase"
+
+    def __hash__(self):
+        return id(self)
+
+    def __init__(self, ctx: PipelineContext, config: MetabaseConfig):
+        super().__init__(ctx)
+        self.config = config
+        self.report = SourceReport()
+
+        login_response = requests.post(
+            f"{self.config.connect_uri}/api/session",
+            None,
+            {
+                "username": self.config.username,
+                "password": self.config.password,
+            },
+        )
+
+        self.access_token = login_response.json()["id"]
+
+        self.session = requests.Session()
+        self.session.headers.update(
+            {
+                "X-Metabase-Session": f"{self.access_token}",
+                "Content-Type": "application/json",
+                "Accept": "*/*",
+            }
+        )
+
+        # Test the connection
+        test_response = self.session.get(f"{self.config.connect_uri}/api/user/current")
+        if test_response.status_code == 200:
+            pass
+
+        if test_response.status_code == 401:
+            self.report.report_failure(
+                key="metabase-session",
+                reason=f"Unable to login for user {self.config.username}. Unauthorized access",
+            )
+
+    def close(self) -> None:
+        response = requests.delete(
+            f"{self.config.connect_uri}/api/session",
+            headers={"X-Metabase-Session": self.access_token},
+        )
+        if response.status_code != 204:
+            self.report.report_failure(
+                key="metabase-session",
+                reason=f"Unable to logout for user {self.config.username}",
+            )
+
+    def emit_dashboard_mces(self) -> Iterable[MetadataWorkUnit]:
+        dashboard_response = self.session.get(
+            f"{self.config.connect_uri}/api/dashboard"
+        )
+        payload = dashboard_response.json()
+        for dashboard_info in payload:
+            dashboard_snapshot = self.construct_dashboard_from_api_data(dashboard_info)
+
+            mce = MetadataChangeEvent(proposedSnapshot=dashboard_snapshot)
+            wu = MetadataWorkUnit(id=dashboard_snapshot.urn, mce=mce)
+            self.report.report_workunit(wu)
+
+            yield wu
+
+    def construct_dashboard_from_api_data(
+        self, dashboard_info: dict
+    ) -> DashboardSnapshot:
+        dashboard_url = (
+            f"{self.config.connect_uri}/api/dashboard/{dashboard_info['id']}"
+        )
+        dashboard_response = self.session.get(dashboard_url)
+        dashboard_details = dashboard_response.json()
+        dashboard_urn = f"urn:li:dashboard:({self.platform},{dashboard_details['id']})"
```

Comment:
would be better to use `builder.make_dashboard_urn`

ditto throughout - there's also make_chart_urn, make_user_urn, make_dataset_urn, etc.

Reply:
+1

Reply:
done. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)